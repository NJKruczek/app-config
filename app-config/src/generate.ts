import { extname, join } from 'path';
import { outputFile } from 'fs-extra';
import { stripIndent } from 'common-tags';
import {
  quicktype,
  RendererOptions,
  JSONSchemaInput,
  JSONSchemaSourceData,
  InputData,
} from 'quicktype-core';
import { loadMetaConfig, Options as MetaOptions } from './meta';
import { loadSchema, JSONSchema, Options as SchemaOptions } from './schema';
import { logger } from './logging';

export interface Options {
  directory?: string;
  schemaOptions?: SchemaOptions;
  metaOptions?: MetaOptions;
}

export interface GenerateFile {
  file: string;

  // Quicktype options
  name?: string;
  type?: string;
  augmentModule?: boolean;
  leadingComments?: string[];
  rendererOptions?: RendererOptions;
}

export async function generateTypeFiles({ directory, schemaOptions, metaOptions }: Options = {}) {
  const { value: schema } = await loadSchema({ directory, ...schemaOptions });
  const {
    value: { generate = [] },
  } = await loadMetaConfig({ directory, ...metaOptions });

  const metaDirectory = metaOptions?.directory ?? directory ?? '.';

  await Promise.all(
    generate.map(
      async ({
        file,
        type = extname(file).slice(1),
        name = 'Config',
        augmentModule = true,
        leadingComments,
        rendererOptions = {},
      }) => {
        logger.info(`Generating ${file} as ${type}`);
        const lines = await generateQuicktype(
          schema,
          type,
          name,
          augmentModule,
          leadingComments,
          rendererOptions,
        );

        await outputFile(join(metaDirectory, file), `${lines.join('\n')}${'\n'}`);
      },
    ),
  );

  return generate;
}

export async function generateQuicktype(
  schema: JSONSchema,
  type: string,
  name: string,
  augmentModule: boolean = true,
  leadingComments: string[] = [
    'AUTO GENERATED CODE',
    "Run app-config with 'generate' command to regenerate this file",
  ],
  rendererOptions: RendererOptions = {},
): Promise<string[]> {
  const src: JSONSchemaSourceData = {
    name,
    schema: JSON.stringify(schema),
  };

  const inputData = new InputData();
  await inputData.addSource('schema', src, () => new JSONSchemaInput(undefined));

  if (['ts', 'typescript', 'flow'].includes(type)) {
    Object.assign(rendererOptions, {
      'just-types': 'true',
      'runtime-typecheck': 'false',
      ...rendererOptions,
    });
  }

  const { lines } = await quicktype({
    inputData,
    lang: type,
    indentation: '  ',
    leadingComments,
    rendererOptions,
  });

  if (['ts', 'typescript'].includes(type)) {
    lines.splice(leadingComments.length, 0, '', "import '@lcdev/app-config';");

    // some configs are empty, so just mark them as an empty object
    if (!lines.some((line) => line.startsWith('export'))) {
      lines.push(`export interface ${name} {}\n`);
    }

    if (augmentModule !== false) {
      lines.push(
        ...[
          '// augment the default export from app-config',
          "declare module '@lcdev/app-config' {",
          `  export interface ExportedConfig extends ${name} {}`,
          '}',
        ],
      );
    }

    return (
      lines
        // this is a fix for quicktype, which adds an Object postfix, sometimes
        .map((line) => line.replace(`interface ${name}Object`, `interface ${name}`))
        .map((line) => line.replace(`: Date`, `: string`))
        .map((line) => line.replace(/: +(\w)/, ': $1'))
    );
  }

  if (['go', 'golang'].includes(type)) {
    const importIndex = lines.findIndex((line) => line.startsWith('import '));

    const imports = [
      'import (',
      '  "encoding/json"',
      '  "log"',
      '  "os"',
      '  "path/filepath"',
      '',
      '  "github.com/xeipuuv/gojsonschema"',
      ')',
    ];

    const singleton = stripIndent`
      var config ${name}

      func init() {
        envValue := os.Getenv("APP_CONFIG")

        if envValue == "" {
          log.Panic("The APP_CONFIG environment variable was undefined")
        }

        loaded, err := UnmarshalConfig([]byte(envValue))

        if err != nil {
          log.Panic("Could not parse APP_CONFIG environment variable: ", err)
        }

        schemaPath, err := filepath.Abs(".app-config.schema.json")

        if err != nil {
          log.Panic("Could not find .app-config.schema.json file: ", err)
        }

        schemaLoader := gojsonschema.NewReferenceLoader("file://" + schemaPath)
        documentLoader := gojsonschema.NewGoLoader(loaded)

        result, err := gojsonschema.Validate(schemaLoader, documentLoader)

        if err != nil {
          log.Panic("Could not validate app-config: ", err.Error())
        }

        if !result.Valid() {
          for _, desc := range result.Errors() {
            log.Printf("- %s\\n", desc)
          }

          log.Panic("The app-config value was not valid.")
        }

        config = loaded
      }

      func GetConfig() ${name} {
        return config
      }
    `.split('\n');

    lines.splice(importIndex, 1, ...imports, '', ...singleton);
    lines.splice(0, 0, '// @generated by app-config', '');

    // get close to gofmt
    return lines.map((line) =>
      line.replace(/ +$/g, '').replace(/^( {2})+/g, (match) => {
        return '\t'.repeat(match.length / 2);
      }),
    );
  }

  return lines;
}
