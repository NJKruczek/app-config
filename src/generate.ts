import { join, basename, extname } from 'path';
import { outputFile } from 'fs-extra';
import { stringify, extToFileType } from './file-loader';
import { loadConfig, ConfigObject } from './config';
import { loadSchema, SchemaRefs } from './schema';
import { loadMeta, resetMetaProps } from './meta';
import * as refParser from 'json-schema-ref-parser';
import {
  quicktype,
  RendererOptions,
  JSONSchema,
  JSONSchemaInput,
  JSONSchemaStore,
  JSONSchemaSourceData,
  InputData,
} from 'quicktype-core';

export interface GenerateFile {
  file: string;
  type?: string;

  // Quicktype options
  name?: string;
  augmentModule?: boolean;
  leadingComments?: string[];
  rendererOptions?: { [key: string]: string };

  // Config output options
  includeSecrets?: boolean;
  select?: string;
}

export const generateTypeFiles = async (cwd = process.cwd()) => {
  resetMetaProps();

  const [{ schema, schemaRefs }, { config, nonSecrets }] = await Promise.all([
    loadSchema(cwd),
    loadConfig(cwd),
  ]);

  const meta = await loadMeta(cwd);

  const { generate = [] } = meta;

  // default to PascalCase with non-word chars removed
  const normalizeName = (file: string) => basename(file, extname(file))
      .split(/[^\w]/)
      .map(s => `${s.charAt(0).toUpperCase()}${s.slice(1)}`)
      .join('');

  await Promise.all(generate.map(async ({
    file,
    type = extname(file).slice(1),
    name = normalizeName(file),
    augmentModule = true,
    leadingComments,
    rendererOptions = {},
    includeSecrets = false,
    select = '#',
  }) => {
    let lines;

    switch (type) {
      case 'json':
      case 'json5':
      case 'toml':
      case 'yml':
      case 'yaml': {
        const output = includeSecrets ? config : nonSecrets;
        const refs = await refParser.resolve(output);
        const selected = refs.get(select);

        lines = [stringify(selected as ConfigObject, extToFileType(type))];
        break;
      }
      default: {
        lines = await generateQuicktype(
          schema, schemaRefs, file, type, name, augmentModule, leadingComments, rendererOptions,
        );
        break;
      }
    }

    await outputFile(join(cwd, file), `${lines.join('\n')}${'\n'}`);
  }));

  return generate;
};

const generateQuicktype = async (
  schema: ConfigObject,
  schemaRefs: SchemaRefs | undefined,
  file: string,
  type: string,
  name: string,
  augmentModule: boolean,
  leadingComments: string[] | undefined,
  rendererOptions: RendererOptions,
) => {
  const src: JSONSchemaSourceData = {
    name,
    schema: JSON.stringify(schema),
  };

  class FetchingJSONSchemaStore extends JSONSchemaStore {
    async fetch(address: string): Promise<JSONSchema | undefined> {
      return (schemaRefs as any)[address];
    }
  }

  const inputData = new InputData();
  await inputData.addSource(
    'schema',
    src,
    () => new JSONSchemaInput(new FetchingJSONSchemaStore(), []),
  );

  const { lines } = await quicktype({
    inputData,
    lang: type,
    indentation: '  ',
    leadingComments: leadingComments || [
      'AUTO GENERATED CODE',
      'Run app-config with \'generate\' command to regenerate this file',
    ],
    rendererOptions: {
      'just-types': 'true',
      'runtime-typecheck': 'false',
      ...rendererOptions,
    },
  });

  if (type === 'ts' && augmentModule !== false) {
    lines.push(...[
      '// augment the default export from app-config',
      "declare module '@servall/app-config' {",
      `  export interface ExportedConfig extends ${name} {}`,
      '}',
    ]);
  }

  return lines;
};
