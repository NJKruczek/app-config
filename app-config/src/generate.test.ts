import { generateQuicktype } from './generate';

describe('TypeScript File Generation', () => {
  it('creates a simple TypeScript file', async () => {
    const schema = {
      type: 'object',
      additionalProperties: false,
      properties: {
        foo: { type: 'string' },
      },
    };

    const generated = await generateQuicktype(schema, 'ts', 'Configuration');

    expect(generated).toMatchSnapshot();
  });
});
