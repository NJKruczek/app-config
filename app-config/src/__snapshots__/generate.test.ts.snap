// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flow File Generation creates a simple flow file 1`] = `
Array [
  "// @flow",
  "",
  "// AUTO GENERATED CODE",
  "// Run app-config with 'generate' command to regenerate this file",
  "",
  "export type Configuration = {",
  "  foo?: string;",
  "};",
  "",
]
`;

exports[`Golang File Generation creates a simple Go file 1`] = `
Array [
  "package main",
  "",
  "import \\"encoding/json\\"",
  "",
  "func UnmarshalConfiguration(data []byte) (Configuration, error) {",
  "  var r Configuration",
  "  err := json.Unmarshal(data, &r)",
  "  return r, err",
  "}",
  "",
  "func (r *Configuration) Marshal() ([]byte, error) {",
  "  return json.Marshal(r)",
  "}",
  "",
  "type Configuration struct {",
  "  Foo *string \`json:\\"foo,omitempty\\"\`",
  "}",
  "",
]
`;

exports[`Rust File Generation creates a simple Rust file 1`] = `
Array [
  "/// AUTO GENERATED CODE",
  "/// Run app-config with 'generate' command to regenerate this file",
  "",
  "extern crate serde_derive;",
  "",
  "#[derive(Serialize, Deserialize)]",
  "pub struct Configuration {",
  "  #[serde(rename = \\"foo\\")]",
  "  foo: Option<String>,",
  "}",
  "",
]
`;

exports[`TypeScript File Generation corrects Date type in TypeScript files 1`] = `
"// AUTO GENERATED CODE
// Run app-config with 'generate' command to regenerate this file

import '@lcdev/app-config';

export interface MyCustomConfigName {
  x?: string;
}

// augment the default export from app-config
declare module '@lcdev/app-config' {
  export interface ExportedConfig extends MyCustomConfigName {}
}
"
`;

exports[`TypeScript File Generation creates a TypeScript file from meta file properties 1`] = `
"// AUTO GENERATED CODE
// Run app-config with 'generate' command to regenerate this file

import '@lcdev/app-config';

export interface MyCustomConfigName {
  x?: number;
}

// augment the default export from app-config
declare module '@lcdev/app-config' {
  export interface ExportedConfig extends MyCustomConfigName {}
}
"
`;

exports[`TypeScript File Generation creates a TypeScript file from schema with many $ref properties 1`] = `
"// AUTO GENERATED CODE
// Run app-config with 'generate' command to regenerate this file

import '@lcdev/app-config';

export interface Config {
  a?: A;
  root?: boolean;
}

export interface A {
  b: B;
}

export interface B {
  c: C;
}

export interface C {
  d: boolean;
}

// augment the default export from app-config
declare module '@lcdev/app-config' {
  export interface ExportedConfig extends Config {}
}
"
`;

exports[`TypeScript File Generation creates a simple TypeScript file 1`] = `
Array [
  "// AUTO GENERATED CODE",
  "// Run app-config with 'generate' command to regenerate this file",
  "",
  "import '@lcdev/app-config';",
  "",
  "export interface Configuration {",
  "  foo?: string;",
  "}",
  "",
  "// augment the default export from app-config",
  "declare module '@lcdev/app-config' {",
  "  export interface ExportedConfig extends Configuration {}",
  "}",
]
`;
